### PR TITLE
Hotfix/wan issue [4322]

### PR DESCRIPTION
**Pull request by JuanCarlos-eProsima**
[#4279]
- Fast RTPS class TCPv4Transport modified to proper WAN operation.
- HelloWorldTCP example WAN enabled.

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1079/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1049/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1063/)

File: examples/C++/HelloWorldExampleTCP/HelloWorldPublisher.cpp [hunk 1]
```diff
@@ -57,7 +57,12 @@ bool HelloWorldPublisher::init()
     descriptor->sendBufferSize = 0;
     descriptor->receiveBufferSize = 0;
     //descriptor->set_WAN_address("127.0.0.1");
-    descriptor->add_listener_port(5100);
+	if (!wan_ip.empty())
```
Comment (MiguelCompany) [inline examples/C++/HelloWorldExampleTCP/HelloWorldPublisher.cpp]: No TABS please

File: examples/C++/HelloWorldExampleTCP/HelloWorldPublisher.h [hunk 2]
```diff
@@ -34,7 +34,7 @@ class HelloWorldPublisher {
 	HelloWorldPublisher();
 	virtual ~HelloWorldPublisher();
 	//!Initialize
-	bool init();
+	bool init(const std::string &wan_ip, unsigned short port);
```
Comment (MiguelCompany) [inline examples/C++/HelloWorldExampleTCP/HelloWorldPublisher.h]: TABS. Take the chance to change the whole file

File: examples/C++/HelloWorldExampleTCP/HelloWorldSubscriber.cpp [hunk 3]
```diff
@@ -36,15 +36,23 @@ mp_subscriber(nullptr)
...
+bool HelloWorldSubscriber::init(const std::string &wan_ip, unsigned short port)
 {
     ParticipantAttributes PParam;
     int32_t kind = LOCATOR_KIND_TCPv4;
 
     Locator_t initial_peer_locator;
     initial_peer_locator.kind = kind;
-    IPLocator::setIPv4(initial_peer_locator, "127.0.0.1");
-    initial_peer_locator.port = 5100;
+	if (!wan_ip.empty())
```
Comment (MiguelCompany) [inline examples/C++/HelloWorldExampleTCP/HelloWorldSubscriber.cpp]: TABS

File: examples/C++/HelloWorldExampleTCP/HelloWorldSubscriber.h [hunk 4]
```diff
@@ -37,7 +37,7 @@ class HelloWorldSubscriber {
 	HelloWorldSubscriber();
 	virtual ~HelloWorldSubscriber();
 	//!Initialize the subscriber
-	bool init();
+	bool init(const std::string &wan_ip, unsigned short port);
```
Comment (MiguelCompany) [inline examples/C++/HelloWorldExampleTCP/HelloWorldSubscriber.h]: TABS. Take the chance to change the whole file

File: examples/C++/HelloWorldExampleTCP/HelloWorld_main.cpp [hunk 5]
```diff
@@ -25,67 +25,93 @@
...
+            {
+                port = atoi(argv[3]);
+            }
         }
     }
     else
     {
         std::cout << "publisher OR subscriber argument needed" << std::endl;
-        std::cout << "publisher has optional arguments: publisher [times] [interval] " << std::endl;
+        std::cout << "publisher has optional arguments: publisher [times] [interval] [wan_ip] [port] " << std::endl;
```
Comment (MiguelCompany) [inline examples/C++/HelloWorldExampleTCP/HelloWorld_main.cpp]: Add info line about subscriber optional arguments.

File: src/cpp/transport/TCPv4Transport.cpp [hunk 6]
```diff
@@ -297,6 +297,32 @@ void TCPv4Transport::EndpointToLocator(const ip::tcp::endpoint& endpoint, Locato
     IPLocator::setIPv4(locator, ipBytes.data());
 }
 
+bool TCPv4Transport::fillMetatrafficUnicastLocator(Locator_t &locator, uint32_t metatraffic_unicast_port) const
+{
+    bool result = TCPTransportInterface::fillMetatrafficUnicastLocator(locator, metatraffic_unicast_port);
+
+    const TCPv4TransportDescriptor* config = static_cast<const TCPv4TransportDescriptor*>(GetConfiguration());
+    if (config->wan_addr[0] != 0 || config->wan_addr[1] != 0 || config->wan_addr[2] != 0 || config->wan_addr[3] != 0)
```
Comment (MiguelCompany) [inline src/cpp/transport/TCPv4Transport.cpp]: I think what should be done is to write the wan address on the wan part of the input locator.

File: src/cpp/transport/TCPv4Transport.cpp [hunk 7]
```diff
@@ -297,6 +297,32 @@ void TCPv4Transport::EndpointToLocator(const ip::tcp::endpoint& endpoint, Locato
...
+
+    return result;
+}
+
+bool TCPv4Transport::fillUnicastLocator(Locator_t &locator, uint32_t well_known_port) const
+{
+    bool result = TCPTransportInterface::fillUnicastLocator(locator, well_known_port);
+
+    const TCPv4TransportDescriptor* config = static_cast<const TCPv4TransportDescriptor*>(GetConfiguration());
+    if (config->wan_addr[0] != 0 || config->wan_addr[1] != 0 || config->wan_addr[2] != 0 || config->wan_addr[3] != 0)
```
Comment (MiguelCompany) [inline src/cpp/transport/TCPv4Transport.cpp]: I think what should be done is to write the wan address on the wan part of the input locator.

File: examples/C++/HelloWorldExampleTCP/HelloWorldPublisher.cpp (hunk 1 above)
Comment (JuanCarlos-eProsima) [reply examples/C++/HelloWorldExampleTCP/HelloWorldPublisher.cpp]: Fixed

File: examples/C++/HelloWorldExampleTCP/HelloWorldPublisher.h (hunk 2 above)
Comment (JuanCarlos-eProsima) [reply examples/C++/HelloWorldExampleTCP/HelloWorldPublisher.h]: Fixed

File: examples/C++/HelloWorldExampleTCP/HelloWorldSubscriber.cpp (hunk 3 above)
Comment (JuanCarlos-eProsima) [reply examples/C++/HelloWorldExampleTCP/HelloWorldSubscriber.cpp]: Fixed

File: examples/C++/HelloWorldExampleTCP/HelloWorldSubscriber.h (hunk 4 above)
Comment (JuanCarlos-eProsima) [reply examples/C++/HelloWorldExampleTCP/HelloWorldSubscriber.h]: Fixed

File: examples/C++/HelloWorldExampleTCP/HelloWorld_main.cpp (hunk 5 above)
Comment (JuanCarlos-eProsima) [reply examples/C++/HelloWorldExampleTCP/HelloWorld_main.cpp]: Updated with a small explanation of the arguments

File: src/cpp/transport/TCPv4Transport.cpp (hunk 7 above)
Comment (JuanCarlos-eProsima) [reply src/cpp/transport/TCPv4Transport.cpp]: We have applied the suggestions to the transport

File: src/cpp/transport/TCPv4Transport.cpp (hunk 6 above)
Comment (JuanCarlos-eProsima) [reply src/cpp/transport/TCPv4Transport.cpp]: We have applied the suggestions to the transport

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1084/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1054/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1068/)

**Pull request merged**